src/osemgrep/reporting/Matches_report.ml

```diff
@@ -293,49 +332,54 @@ let pp_text_outputs ~max_chars_per_line ~max_lines_per_finding ~color_output ppf
+  |> Common.group_by (fun (m : Out.cli_match) ->
+         match product_of_cli_match m with
+         | `SCA ->
+             (* TO PORT:
+                                  subgroup = match.exposure_type or "undetermined"
 
-              figuring out the product, python uses (rule.py):
-                 RuleProduct.sca
-                 if "r2c-internal-project-depends-on" in self._raw
-                 else RuleProduct.sast
+                   figuring out the product, python uses (rule.py):
+                      RuleProduct.sca
+                      if "r2c-internal-project-depends-on" in self._raw
+                      else RuleProduct.sast
 
-              and exposure_type (rule_match.py):
-              if "sca_info" not in self.extra:
-                  return None
+                   and exposure_type (rule_match.py):
+                   if "sca_info" not in self.extra:
+                       return None
 
-              if self.metadata.get("sca-kind") == "upgrade-only":
-                  return "reachable"
-              elif self.metadata.get("sca-kind") == "legacy":
-                  return "undetermined"
-              else:
-                  return "reachable" if self.extra["sca_info"].reachable else "unreachable"
-           *)
-           if is_blocking m.Out.extra.Out.metadata then `Blocking
-           else `Nonblocking)
-  in
-  (* if not is_ci_invocation *)
-  let groups =
-    let merged =
-      (try List.assoc `Nonblocking groups with
-      | Not_found -> [])
-      @
-      try List.assoc `Blocking groups with
-      | Not_found -> []
-    in
-    (`Merged, merged)
-    :: List.filter
-         (fun (k, _) -> not (k = `Nonblocking || k = `Blocking))
-         groups
-  in
-  groups
+                   if self.metadata.get("sca-kind") == "upgrade-only":
+                       return "reachable"
+                   elif self.metadata.get("sca-kind") == "legacy":
+                       return "undetermined"
+                   else:
+                       return "reachable" if self.extra["sca_info"].reachable else "unreachable" *)
+             `Undetermined
+         | `SAST when is_blocking m.Out.extra.Out.metadata -> `Blocking
+         | `SAST -> `Nonblocking
+         | `Secrets -> (
+             match m.extra.validation_state with
+             | Some `Confirmed_valid -> `Valid
+             | Some `Confirmed_invalid -> `Invalid
+             | Some `Validation_error -> `Validation_error
+             | Some `No_validator
+             | None ->
+                 `Unvalidated))
+  |> (fun groups ->
+       (* TO PORT:
+          if not is_ci_invocation: *)
+       let merged =
+         (try List.assoc `Nonblocking groups with
+         | Not_found -> [])
+         @
+         try List.assoc `Blocking groups with
+         | Not_found -> []
+       in
+       (`Merged, merged)
+       :: List.filter
+            (fun (k, _) -> not (k = `Nonblocking || k = `Blocking))
```

Comment:
Could we maybe do a partition instead to avoid needing quite so many list traversals? I suppose this isn't that long so it isn't really an issue but it does seem a bit silly to separately filter these.

Reply:
I think it is conceptually simpler to just do more list traversals and the bounds is a relatively small constant. I feel like this would be over-optimizing.